### PR TITLE
test(negative): implement backupstore setup

### PR DESCRIPTION
**Pull request by yangchiu**
test: implement backupstore setup

(1) add backupstore.py for robot test cases
(2) only support s3 now, the subprocess parts in backupstore.py need to be refined to make nfs work
(3) fix wrong longorn client url issue when using it out-of-cluster (the same issue as https://github.com/longhorn/longhorn-tests/pull/1300#discussion_r1273250265)

For https://github.com/longhorn/longhorn/issues/6710

Signed-off-by: Yang Chiu <yang.chiu@suse.com>

### CODE REVIEW

**Review by innobead — CHANGES_REQUESTED** [review]
Some feedbacks, but I didn't review all of them. Some needs to resolve first.

backupstore.py seems verbose and unstructured. I would suggest doing below.

Use different python modules to aggregate the related codes.
- backupstore/__init__.py including all common functions for backupstore
   backupstore/nfs.py, minio.py
 
 Also, using the common naming (verb) for the function and method name instead, so `backupstore_cleanup` should be changed to `cleanup_backupstore` and add it to backupstore/__init__.py. 
 
 Please review the code and do corresponding update.

File: e2e/libs/backupstore.py
Comment (yangchiu) [reply e2e/libs/backupstore.py]: This can be refined later if it actually has the situation that deleting backup volumes is unexpected time-consuming. If it's observed, it can be refined.

File: e2e/libs/backupstore.py
Comment (yangchiu) [reply e2e/libs/backupstore.py]: Moved these backup specific values to a backup class which contains all backup specific code.

File: e2e/libs/backupstore.py
Comment (yangchiu) [reply e2e/libs/backupstore.py]: Removed.

File: e2e/libs/backupstore.py
Comment (yangchiu) [reply e2e/libs/backupstore.py]: Refined.

File: e2e/libs/backupstore.py
Comment (yangchiu) [reply e2e/libs/backupstore.py]: Refined.

File: e2e/libs/backupstore.py
Comment (yangchiu) [reply e2e/libs/backupstore.py]: Refined.

**Comment by yangchiu** [comment]
> Some feedback, but I didn't review all of them. Some need to be resolved first.
> 
> backupstore.py seems verbose and unstructured. I would suggest doing the below.
> 
> Use different Python modules to aggregate the related codes.
> 
>     * backupstore/__init__.py including all common functions for backupstore
>       backupstore/nfs.py
>       backupstore/minio.py (but actually it's s3)
> 
> 
> Also, using the common naming (verb) for the function and method name instead, `backupstore_cleanup` should be changed to `cleanup_backupstore` and added to backupstore/__init__.py.
> 
> Please review the code and do the corresponding update.

Refined to have class structure.

**Comment by yangchiu** [comment]
@longhorn/qa @innobead This PR is required to run negative test case `Reboot Volume Node While Heavy Writing And Recurring Jobs Exist` in v1.4.4 release test plan.

File: e2e/libs/backupstore/base.py
Comment (c3y1huang) [inline e2e/libs/backupstore/base.py:165]:
```suggestion
    def cleanup_backup_volumes(self, client):
```

File: e2e/libs/backupstore/minio.py
Comment (c3y1huang) [inline e2e/libs/backupstore/minio.py:94]: Where is TEMP_FILE_PATH defined?

**Review by c3y1huang — CHANGES_REQUESTED** [review]

File: e2e/libs/backupstore/base.py
Comment (yangchiu) [reply e2e/libs/backupstore/base.py:40]: Can be done later when we have setting related test cases since currently backupstore setting is the only setting exists in the test code.

File: e2e/libs/backupstore/minio.py
Comment (yangchiu) [reply e2e/libs/backupstore/minio.py:9]: Not sure about this. Even though the class name is changed to `S3`, the apis called in this class are still from `minio`.

File: e2e/libs/backupstore/minio.py
Comment (yangchiu) [reply e2e/libs/backupstore/minio.py:96]: Refined.

File: e2e/libs/backupstore/base.py
Comment (yangchiu) [reply e2e/libs/backupstore/base.py]: Refined.

File: e2e/keywords/common.resource
Comment (yangchiu) [reply e2e/keywords/common.resource:26]: Put it here to implicitly setup backupstore.

**Review by c3y1huang — DISMISSED** [review]
I assume most of the codes are moved from the integration tests as I see there are some variables that are still undefined, and the functions could be better organized. However, as long as the tests can run I think it's fine to merge this first and enhance later.

cc @innobead

**Comment by yangchiu** [comment]
Need to refine:
1. Rebooting volume node will cause the volume fast detaching and reattaching again **at the moment of the node back to normal**, currently the test case is unable to catch the timing and causes assertion failed.
2. If the node of minio service is rebooted, it's unable to create backup too, and causes the test case failed.

**Comment by yangchiu** [comment]
>     1. Rebooting volume node will cause the volume fast detaching and reattaching again **at the moment of the node back to normal**, currently the test case is unable to catch the timing and causes assertion failed.

Fixed by rearrange the orders of status checking logic.

>     2. If the node of minio service is rebooted, it's unable to create backup too, and causes the test case failed.

Fixed by creating a deployment for minio instead of a plain pod.

And  according to https://github.com/longhorn/longhorn/issues/7956:

> Pushing a backup and a purge once a minute may be a little aggressive.
> Maybe the test can be slightly modified to recur less often.

Refined the recurring job schedule to every 2 minutes.

**Comment by yangchiu** [comment]
@innobead Is it possible to merge this PR? Since the test cases have been listed in release testing plans and they are always blocked by this PR.

**Comment by innobead** [comment]
@yangchiu sounds good. @c3y1huang Please help review this. After that, we can merge this PR.

**Comment by yangchiu** [comment]
@longhorn/qa Please help review this. Many thanks.

File: e2e/libs/backupstore/__init__.py
Comment (c3y1huang) [inline e2e/libs/backupstore/__init__.py:2]: GitHub is showing a red marker (No newline at end of file) for some files in the PR. This is probably not harmful, but it can be rather bothersome. Suggest adding a new line at the end of those files to resolve it.

**Review by c3y1huang — CHANGES_REQUESTED** [review]

File: e2e/libs/recurringjob/recurringjob.py
Comment (yangchiu) [reply e2e/libs/recurringjob/recurringjob.py:22]: Please refer to https://github.com/longhorn/longhorn-tests/pull/1539#issuecomment-1951537233

File: e2e/libs/recurringjob/rest.py
Comment (yangchiu) [reply e2e/libs/recurringjob/rest.py:184]: One thing needs to be mentioned is longhorn client itself uses naming like create_recurring_job‎ and by_id_recurring_job, even though this changes to single word, there is still inconsistency in the file.

File: e2e/libs/backupstore/base.py
Comment (yangchiu) [reply e2e/libs/backupstore/base.py:40]: Sure. I'll organize them in the later PR.

File: e2e/libs/backupstore/base.py
Comment (yangchiu) [reply e2e/libs/backupstore/base.py:14]: I'm still not sure about this. WDYT? @longhorn/qa

File: e2e/libs/recurringjob/rest.py
Comment (c3y1huang) [reply e2e/libs/recurringjob/rest.py:184]: I think that it's fine since it's external to the longhorn-test repo. I understand there are some inconsistencies in the longhorn REST client. Let's just focus on the e2e test codebase now.

File: e2e/libs/backupstore/base.py
Comment (c3y1huang) [reply e2e/libs/backupstore/base.py:14]: As the test grows, it often accumulates numerous constants, like [this](https://github.com/longhorn/longhorn-tests/blob/master/manager/integration/tests/common.py#L30-L302). I believe it would be more organized to manage these contents is dedicated 
 constant file.
 
However, if the team decides not to take this approach, I will understand that decision. As such, we then need to take out the constants in the constant.py files and place them in the relevant places.

File: e2e/libs/keywords/backupstore_keywords.py
Comment (yangchiu) [reply e2e/libs/keywords/backupstore_keywords.py]: Refined. Thanks!

File: e2e/libs/backupstore/base.py
Comment (yangchiu) [reply e2e/libs/backupstore/base.py:14]: It's because `common.py` is written in C-like style and mixing everything together. We can use a more OOP way to handle them. And I think it is reasonable for constants placed in different files if they are specific to and only used by a particular class.

File: e2e/libs/recurringjob/recurringjob.py
Comment (yangchiu) [reply e2e/libs/recurringjob/recurringjob.py:25]: Refined. Thanks!

File: e2e/libs/recurringjob/rest.py
Comment (yangchiu) [reply e2e/libs/recurringjob/rest.py]: Refined. Thanks!

File: e2e/libs/backupstore/__init__.py
Comment (yangchiu) [reply e2e/libs/backupstore/__init__.py]: Refined. Thanks!

File: e2e/libs/backupstore/base.py
Comment (c3y1huang) [reply e2e/libs/backupstore/base.py:40]: Given that we are introducing this new function, I don't see any reason we shouldn't organize this now rather than deferring to a later PR.

File: e2e/libs/keywords/backupstore_keywords.py
Comment (c3y1huang) [inline e2e/libs/keywords/backupstore_keywords.py:55]: Add line break at EOF. Please also check other files.

**Review by c3y1huang — CHANGES_REQUESTED** [review]

File: e2e/libs/backupstore/base.py
Comment (c3y1huang) [reply e2e/libs/backupstore/base.py:14]: I personally prefer the constants in a separate file for easier management, but I am flexible if the team decides otherwise. I believe what's most important is making sure that @longhorn/qa agrees and understands the final decision and that we maintain a consistent convention in the e2e.

Comment (roger-ryao) [reply e2e/libs/backupstore/base.py:14]: Based on my previous experience, my habit is similar to @c3y1huang's suggested usage approach, where I centralize the use of variables in the same variable file, `constants.py`. In the future, when the data in the variable file becomes extensive, it will be easier for us to organize and categorize them further.
However, I currently don't have a concept of the method @yangchiu mentioned: `We can use a more OOP way to handle them.`

File: e2e/libs/keywords/backupstore_keywords.py
Comment (yangchiu) [reply e2e/libs/keywords/backupstore_keywords.py]: Refined. Thank you!

File: e2e/libs/backupstore/base.py
Comment (yangchiu) [reply e2e/libs/backupstore/base.py]: Refined. Thank you!

File: e2e/libs/backupstore/base.py
Comment (yangchiu) [reply e2e/libs/backupstore/base.py]: Refined. Thank you!

File: e2e/libs/backupstore/base.py
Comment (c3y1huang) [inline e2e/libs/backupstore/base.py:33]: `SETTING_BACKUP_TARGET` is not defined. Please check other places.

File: e2e/libs/backupstore/minio.py
Comment (c3y1huang) [inline e2e/libs/backupstore/minio.py:19]: I couldn't find where this is used. Consider Removing if not required.

File: e2e/libs/recurringjob/rest.py
Comment (c3y1huang) [inline e2e/libs/recurringjob/rest.py:124]: Could you explain the purpose of this log?

File: e2e/libs/utility/utility.py
Comment (c3y1huang) [inline e2e/libs/utility/utility.py:192]: This looks like it belongs in the backupstore lib.

**Review by c3y1huang — DISMISSED** [review]

File: e2e/libs/utility/utility.py
Comment (yangchiu) [reply e2e/libs/utility/utility.py]: Refined. Thank you!

File: e2e/libs/recurringjob/rest.py
Comment (yangchiu) [reply e2e/libs/recurringjob/rest.py]: Removed. Thank you!

File: e2e/libs/backupstore/minio.py
Comment (yangchiu) [reply e2e/libs/backupstore/minio.py]: Removed. Thank you!

File: e2e/libs/backupstore/base.py
Comment (yangchiu) [reply e2e/libs/backupstore/base.py]: Refined. Thank you!

**Review by c3y1huang — DISMISSED** [review]

File: e2e/libs/backupstore/minio.py
Comment (yangchiu) [reply e2e/libs/backupstore/minio.py]: Refined. Thank you!

**Review by c3y1huang — APPROVED** [review]

**Pull request merged**
